polkadot/node/network/collator-protocol/src/validator_side_experimental/collation_manager/mod.rs

```diff
@@ -0,0 +1,1177 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::{
+	validator_side::{
+		descriptor_version_sanity_check, error::SecondingError, request_persisted_validation_data,
+		request_prospective_validation_data, BlockedCollationId, PerLeafClaimQueueState,
+	},
+	validator_side_experimental::{
+		common::{
+			Advertisement, CanSecond, CollationFetchError, CollationFetchResponse,
+			ProspectiveCandidate, Score, SecondingRejectionInfo, FAILED_FETCH_SLASH,
+			INSTANT_FETCH_REP_THRESHOLD, UNDER_THRESHOLD_FETCH_DELAY,
+		},
+		error::{Error, FatalResult, Result},
+	},
+	LOG_TARGET,
+};
+use fatality::Split;
+use futures::{channel::oneshot, stream::FusedStream};
+use polkadot_node_network_protocol::{
+	request_response::{outgoing::RequestError, v2 as request_v2, Requests},
+	OurView, PeerId,
+};
+use polkadot_node_primitives::PoV;
+use polkadot_node_subsystem::{
+	messages::{CanSecondRequest, CandidateBackingMessage},
+	ActivatedLeaf, CollatorProtocolSenderTrait,
+};
+use polkadot_node_subsystem_util::{
+	backing_implicit_view::View as ImplicitView, request_claim_queue, request_node_features,
+	request_session_index_for_child, request_validator_groups, request_validators,
+	runtime::recv_runtime,
+};
+use polkadot_primitives::{
+	node_features, CandidateHash, CandidateReceiptV2 as CandidateReceipt, CoreIndex, GroupIndex,
+	GroupRotationInfo, Hash, HeadData, Id as ParaId, PersistedValidationData, SessionIndex,
+};
+use requests::PendingRequests;
+use schnellru::{ByLength, LruMap};
+use sp_keystore::KeystorePtr;
+use std::{
+	collections::{BTreeSet, HashMap, HashSet, VecDeque},
+	time::Instant,
+};
+
+mod requests;
+
+/// Reason for rejecting an advertisement.
+#[derive(Debug, thiserror::Error)]
+pub enum AdvertisementError {
+	#[error("Validator is not assigned to this paraid")]
+	InvalidAssignment,
+	#[error("Duplicate advertisement")]
+	Duplicate,
+	#[error("Advertised relay parent is out of our view")]
+	OutOfOurView,
+	#[error("Peer reached the candidate limit")]
+	PeerLimitReached,
+	#[error("Seconding not allowed by backing subsystem")]
+	BlockedByBacking,
+	#[error("V1 advertisements are only allowed on active leaves")]
+	V1AdvertisementForImplicitParent,
+}
+
+/// Fetched collation data.
+#[derive(Debug, Clone)]
+struct FetchedCollation {
+	/// Candidate receipt.
+	pub candidate_receipt: CandidateReceipt,
+	/// Proof of validity.
+	pub pov: PoV,
+	/// Optional parachain parent head data. This is needed for elastic scaling to work.
+	pub maybe_parent_head_data: Option<HeadData>,
+	/// Optional parent head data hash. This is needed for async backing to work (sent by v2
+	/// protocol).
+	pub maybe_parent_head_data_hash: Option<Hash>,
```

Comment:
Should `parent_head_data_hash` be equal to `parent_head_data.hash()` if they are both `Some` ? Asking because if so, it might be better to handle them with an enum. If not, probably we should rename them ?